src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs

```diff
@@ -966,6 +979,7 @@ protected override void ProcessRecord()
                 }
 
                 _processList.Add(process);
+                _initialNumberOfProcesses += 1;
```

Comment:
I still think we have no need the counter. Below in EndProcessing we enumerate these processes and should check HasExited for every process and finish if HasExited and Any is true.

Reply:
Of the two places I used it within `EndProcessing`, the first one definitely wasn't necessary; `if (_numberOfProcessesToWaitFor > 0)` works just fine.  

I'm not sure what else to do about the `foreach` loop after that, though.  If I understand correctly, as it was written, that block assumed that, if we've reached it, either all processes have exited, or their timeout has elapsed, so any process still running definitely needs a timeout error.  With `-Any`, that logic leads to timeout errors being written for whatever's still running after the first process exits, regardless of whether the timeout has elapsed—or been set at all.  So far, the best alternative I can think of for the conditions I initially put there is replacing `_initialNumberOfProcesses` with `_processList.Count`, which at least gets rid of the extra counter.  Am I overlooking some way to directly check whether a timeout on `_waitHandle` has elapsed?

Reply:
1. We have no need to change existing logic (for waiting all)
2. For Any it makes no sense to analyze process statuses and write any messages - if waiting has terminated for some reason the cmdlet should terminate too.

Reply:
re: your second point, are you saying that, when using `-Timeout` with `-Any`, no error should be written if the timeout elapses?

To make sure we're on the same page, the code I was referring to in my last reply was:

```csharp
foreach (Process process in _processList)
{
    try
    {
        if (!process.HasExited)
        {
            string message = StringUtil.Format(ProcessResources.ProcessNotTerminated, new object[] { process.ProcessName, process.Id });
            ErrorRecord errorRecord = new(new TimeoutException(message), "ProcessNotTerminated", ErrorCategory.CloseError, process);
            WriteError(errorRecord);
        }
    }
    catch (Win32Exception exception)
    {
        WriteNonTerminatingError(process, exception, ProcessResources.ProcessIsNotTerminated, "ProcessNotTerminated", ErrorCategory.CloseError);
    }
}
```

Leaving that code as-is, here's what I'm seeing:

- **With no parameters set:** works fine, as you say.  Without a timeout, this code is only ever reached after all processes have exited, so no errors are written.
- **With `-Timeout` but not `-Any`:** again, works fine.  Either
  - All processes have exited, so no errors are written, or
  -  The timeout has elapsed.  For each process that hasn't exited yet, an error is written saying that that process did not stop during the timeout period.
- **With `-Any` but not `-Timeout`:** we reach this code once a single process has exited.  For each process that hasn't exited yet, an error is written saying that that process did not stop during the timeout period, _even though we didn't set a timeout_.
- **With `-Any` and `-Timeout`:** Either
  - No processes have exited before the timeout period, so a timeout error is written for each process, or
  - We reach this code once one process has exited.  A timeout error is written for each process that hasn't exited, _even though the timeout has not elapsed_.

If we don't care about writing an error when the timeout elapses when using `-Any`, then yeah, wrapping that block with `if (!Any)` gets the job done.  Otherwise, I think we need some way to distinguish between reaching that block because the timeout has elapsed and reaching it because we're using `-Any` and a process has exited.  The way that occurred to me was comparing the number of processes we're still waiting for to the number of processes we started with, but I'm not married to that solution.

Reply:
Oof, I've really been overthinking this.  Thanks for being patient while I missed the forest for the trees.  As you suggested, I've removed `_initialNumberOfProcesses` and made the message-writing code contingent on the return value from `_waitHandle.WaitOne(_timeout * 1000)`.  It looks like this achieves the desired behavior without having to refactor to use `WaitAny()`.